### ISSUE TITLE
It'd be nice to move the compatibility matrix into another package.

### ISSUE THREAD

**Issue by KOLANICH** [FHPythonUtils/LicenseCheck#3]
## Feature
- [x] I have read the comment above and have completed each step
- [x] I have filled in each heading below

### Is your feature request related to a problem? Please describe
I don't need the most of your machinery, I gonna make an own one. But I don't want to maintain an own license compatibility database.

### Describe the solution you'd like
The db to be detached from the machinery.

### Describe alternatives you've considered
Cooperate and redesign this package according to my ideas. But it would likely take long.

### Additional context
I am trying to make a library computing a license of a piece of software automatically combined of other pieces of software with own licenses.

**Comment by FredHappyface** [comment]
I agree with this, should probably have the matrix as some sort of json file or something. Any ideas how this would be best achieved?

This would have the benefit of being language independent too

One idea may be using an `incompatible_licenses.json`

```jsonc
{
    "unlicense": ["mit", "boost", "bsd", ...],
    "public": ["mit", "boost", "bsd", ...],
    ...
}
```

Alternatively, the exact inverse could be used `compatible_licenses.json`

```jsonc
{
    "unlicense": ["unlicense", "public"],
    "public": ["unlicense", "public"],
    ...
}
```

**Comment by KOLANICH** [comment]
I originally thought about
1. fetching the licenses info from GitHub/TLDRLegal via API. They have them annotated with binary features like "may reuse", "may relicense", "must include license".
https://api.github.com/licenses/mit
https://api.github.com/licenses/lgpl-3.0
https://api.github.com/licenses/agpl-3.0
https://tldrlegal.com/api/license/52e58cac81519e003b0000df

2. From these descriptions we can infer a class of a license: public-domain, permissive, copyleft or completely proprietary.

3. Permissive licenses are compatible to each other. So we only need to store explicit compatibily info for viral ones. Copyleft licenses not compatible to each other form families.

4. Then we can apply a some form of a very lightweight domain-specific symbolic solver to detect a family of each license. A family is a cluster of licenses that are not mutually compatible to other clusters. From these clusters and their relations we can build an oriented graph.

5. In this oriented graph we can find clusters that are sinks. These are the licenses the combined work must be licensed under.

6. We can check if these sinks are compatible to each other by the features got via API. If they are not, it may be a copyright violation to make a combined work.

7. If they are compatible, the resulting end license is an SPDX expression joining them with AND.

**Comment by FredHappyface** [comment]
I think that using tldrlegal is probably the best idea there. Can create some script to populate some db/ json file with this basic info and the type of license

2. Can probably regex for the type
3. Are all permissive licenses compatible? A quick search online suggests yes
4. Can you elaborate a little on this? Some of the terminology is a little outside of my knowledge 
5. Sounds like a good approach, not entirely sure how to implement this though
6. Yeah, though I definitely think we should fetch these ahead of time 
7. Would this not be an issue with many copyleft licenses as they are often same-license? - though I guess this would make them incompatible

**Comment by KOLANICH** [comment]
>Can you elaborate a little on this? Some of the terminology is a little outside of my knowledge

Under a `symbolic solver` just a lib solving systems of equations and returning an expression is assummed. I. e. `sympy`. If you have ever worked with Wolfram Mathematica, you would understand what I mean. You give the lib an equation, it returns a solution.

Or we may even not need a symbolic solver, but just an SMT solver like `z3`. SMT solvers output not symbolic expressions, but some solutions. But we have a graph with binary weights, so an SMT solver may fit, and will likely perform much better, than sympy.

But I guess we don't need an SMT solver at all, I feel we just need to find all the strongly connected components.

**Comment by FredHappyface** [comment]
I've made a quick script to scape the endpoint (I'll add some kind of limit to avoid going over any limits)

Would you say this contains all the raw data you need?

```json
{
	"mit-license": {
		"title": "MIT License (Expat)",
		"short": "mit",
		"tags": [
			"Open Source",
			"OSI-Approved",
			"Permissive"
		],
		"must": [
			"Include Copyright",
			"Include License"
		],
		"cannot": [
			"Hold Liable"
		],
		"can": [
			"Commercial Use",
			"Modify",
			"Distribute",
			"Sublicense",
			"Private Use"
		],
		"type": "Permissive"
	}
}
```

**Comment by KOLANICH** [comment]
>Just come to my attention that this breaches https://tldrlegal.com/pages/legal so that's the end of that

Looks like bullshit (though I am not a lawer):
* this notice doesn't prevent web spiders (Google, Yandex, Bing) from scraping and indexing the website, and the website doesn't have `robots.txt`, so they really have taken no measures to prevent the access to their website (that may be prohibited by the ToS) that is very likely to occur completely automatically by dumb bots that cannot read ToSes;
* they have publicly available API working without an API key;
* they have links to JSON on licenses pages;
* there are quite some apps on GH accessing their website automatically (BTW, what is automatically?);

Also facts are not copyrightable, and databases are not covered by copyright, there is a special kind of monopoly for DBs that is weaker than copyright (but it is a shame that this shit exists at all! TBH, IMHO copyrights and all the monopolies must be abolished).

But I feel like GH API is more easier to use:
* they don't use guids for everything and no additional requests is needed
* In fact the dataset they use for the API is available as a part of a git repo, no need to scrape the API and avoid interaction limits, everything is here: https://github.com/github/choosealicense.com/tree/gh-pages/_licenses

I guess if we ever gonna use TLDRLegal, we would have to write a converter into GH data model. So, first I propose to write some abstractions about the licenses and a lib for getting qnd using license info in an OOP way, and a CLI tool that would just print the info about the license by its SPDX id into a console (feel free to use my RichConsole for coloring and styling the text). 
Having a CLI tool won't replace tests, but it would be convenient for debugging and designing the lib to have it.

The next step is to improve the lib to derive licenses kinds from features and the tool to show them. ML can be used here, but IMHO it is completely not needed here and I guess a few handcoded rules would be much better (easy, small, understandable and debuggable) here.

>Would you say this contains all the raw data you need?

Initially - yes (but as I have said, I'd prefer to have the same form for the data fetched from GH and from TL;DRLegal). It even contains a license class. Also TL;DRLegal API provides the info about previous versions of the license with the same name, may make sense to try do discover explicit compatibility from analysis of the text for marker expressions like "any later version".

**Comment by FredHappyface** [comment]
Regarding the first point: I still would like to obtain permission before using the API which has been kindly given. Though it perhaps would be useful for tldrlegal to implement some of what you've suggested. In the meantime I've added a sleep function to the scraper to help control the load on the server (I suspect this is a pretty small amount of the traffic at a given point - let's keep it that way).

I personally do not disagree with the concept of copyright - but that may be something we agree to disagree on 😉

The github API seems a bit limited IMO. Whilst it has some spdx ids, it has a tiny range of licenses (I think it was something like 30?) Plus its harder to identify the license type (Permissive, Copyleft etc) 

This is what I have so far: https://github.com/FHPythonUtils/LicenseMatrix 

There's some use of SequenceMatcher for where strings are off a bit. I've found this to be ok so far. There's also some naive compatibility checkers in the license class.
